File: day21/day21-1.mjs
import * as fs from "node:fs";
import _ from "lodash";

function readLines() {
  return fs.readFileSync("input.txt", "utf-8").split(/\n/);
}

const monkeys = new Map();
readLines().forEach((line) => {
  const [name, rest] = line.split(": ");
  const args = rest.split(" ");
  monkeys.set(name, args);
});

console.log("INPUT:", monkeys);

function getResult(name) {
  const op = monkeys.get(name);
  if (op.length === 1) {
    return parseInt(op[0], 10);
  }

  if (op[1] === "+") {
    return getResult(op[0]) + getResult(op[2]);
  }
  if (op[1] === "-") {
    return getResult(op[0]) - getResult(op[2]);
  }
  if (op[1] === "*") {
    return getResult(op[0]) * getResult(op[2]);
  }
  if (op[1] === "/") {
    return getResult(op[0]) / getResult(op[2]);
  }
}

const result = getResult("root");
console.log("RESULT:", result);

File: day21/day21-1.py
def parseLine(line):
    parts = line.split(': ')
    return (parts[0], parts[1].split(' '))


with open("input.txt") as f:
    input = f.read()

lines = input.split('\n')
nodes = dict(parseLine(line) for line in lines)
print("Input: ", nodes)


def processNode(key):
    node = nodes[key]
    if len(node) == 1:
        return int(node[0])

    left = processNode(node[0])
    right = processNode(node[2])

    match node[1]:
        case '+':
            return left + right
        case '-':
            return left - right
        case '*':
            return left * right
        case '/':
            return left / right
        case _:
            raise Exception("Unknown operator: " + node[1])


print("Result: ", processNode("root"))
